etcdserver/security/security.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package security
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"sort"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/crypto/bcrypt"
+	"github.com/coreos/etcd/etcdserver"
+	"github.com/coreos/etcd/etcdserver/etcdhttp/httptypes"
+	"github.com/coreos/etcd/pkg/types"
+)
+
+const (
+	StorePermsPrefix = "/2"
+)
+
+type Store struct {
+	server  *etcdserver.EtcdServer
+	timeout time.Duration
+}
+
+type User struct {
+	User     string    `json:"user"`
+	Password string    `json:"password,omitempty"`
+	Roles    []string  `json:"roles"`
+	Grant    []string  `json:"grant,omitempty"`
+	Revoke   []string  `json:"revoke,omitempty"`
+	Modified time.Time `json:"lastModified"`
+}
+
+type Role struct {
+	Role        string      `json:"role"`
+	Permissions Permissions `json:"permissions"`
+	Grant       Permissions `json:"grant"`
+	Revoke      Permissions `json:"revoke"`
+	Modified    time.Time   `json:"lastModified"`
+}
+
+type Permissions struct {
+	KV rwPermission `json:"kv"`
+}
+
+type rwPermission struct {
+	Read  []string `json:"read"`
+	Write []string `json:"write"`
+}
+
+func mergeErr(s string, v ...interface{}) *httptypes.HTTPError {
+	return httptypes.NewHTTPError(400, fmt.Sprintf(s, v...))
+}
+
+func NewStore(server *etcdserver.EtcdServer, timeout time.Duration) *Store {
+	s := &Store{
+		server:  server,
```

Comment:
we only need the do function. can we create an interface instead of passing etcdserver around?